src/app/add-item/add-item.component.spec.ts

```diff
@@ -0,0 +1,168 @@
+    expect(nameField.valid).toBeFalsy();
+  });
+
+  it('name field validity', () => {
+    let errors = {};
+    errors = nameField.errors || {};
+    nameField.setValue('');
+    expect(nameField.hasError('required')).toBeTruthy();
+  });
+
+  it('projectName field invalid when empty', () => {
+    expect(projectField.valid).toBeFalsy();
+  });
+
+  it('projectName field validity', () => {
+    let errors = {};
+    errors = projectField.errors || {};
+    projectField.setValue('');
+    expect(projectField.hasError('required')).toBeTruthy();
+  });
+
+  it('dueDate field valid when empty', () => {
+    const dueDateField = component.dialogForm.controls.dueDate;
+    expect(dueDateField.valid).toBeTruthy();
+  });
+
+  it('description field valid when empty', () => {
+    const descriptionField = component.dialogForm.controls.description;
+    expect(descriptionField.valid).toBeTruthy();
+  });
+
+  it('form valid when nameField and projectField filled', () => {
+    nameField.setValue('test');
+    projectField.setValue('test');
+    expect(component.dialogForm.valid).toBeTruthy();
+  });
+
+  it('button disabled when nameField and projectField empty', () => {
+    const createButton = fixture.debugElement.query(By.css('.create'));
+    fixture.detectChanges();
+    expect(createButton.nativeElement.disabled).toBeTruthy();
+  });
+
+  it('button active when nameField and projectField filled', () => {
+    const createButton = fixture.debugElement.query(By.css('.create'));
+    nameField.setValue('test');
+    projectField.setValue('test');
+    fixture.detectChanges();
+    expect(createButton.nativeElement.disabled).toBeFalsy();
+  });
+
+  it('on createActionItem() form disabled', () => {
+    expect(component.dialogForm.valid).toBeFalsy();
+    nameField.setValue('test name');
+    projectField.setValue('test project');
+    expect(component.dialogForm.valid).toBeTruthy();
+    component.createActionItem();
+    expect(nameField.enabled).toBe(false);
+    expect(projectField.enabled).toBe(false);
+    expect(dueDateField.enabled).toBe(false);
+    expect(descriptionField.enabled).toBe(false);
+  });
+
+  // it('createActionItem() should call close()', () => {
+  //   const spyObj = spyOn(component.dialogRef, 'close').and.callThrough();
+  //   component.createActionItem();
+  //   fixture.detectChanges();
+  //   expect(spyObj).toHaveBeenCalled();
+  // });
+
+  it('should call close() when close button clicked', () => {
+    spyOn(component, 'close');
+    const cancelButton: DebugElement = fixture.debugElement.query(By.css('.close'));
+    cancelButton.triggerEventHandler('click', null);
+    fixture.detectChanges();
+    expect(component.close).toHaveBeenCalled();
+  });
+
+  it('should display spinner when createActionItem() called and deactivate it after 3s', done => {
+    component.createActionItem();
+    setTimeout(() => {
+      expect(component.buttonSpinner).toBe(true);
+      done();
+    }, 101);
+    setTimeout(() => {
```

Comment:
Please read about `fakeAsync` and `thick` functions for testing code with timeouts in Angular.